cmd/snap-update-ns/utils.go

```diff
@@ -184,6 +224,38 @@ func secureMkdirAll(name string, perm os.FileMode, uid, gid int) error {
 	return nil
 }
 
+// secureMkfileAll is a secure implementation of "mkdir -p $(dirname $1) && touch $1".
+//
+// This function is like secureMkdirAll but it creates an empty file instead of
+// a directory for the final path component. Each created directory component
+// is chowned to the desired user and group.
+func secureMkfileAll(name string, perm os.FileMode, uid, gid int) error {
+	logger.Debugf("secure-mkfile-all %q %q %d %d", name, perm, uid, gid)
+
+	// Only support absolute paths to avoid bugs in snap-confine when
+	// called from anywhere.
+	if !filepath.IsAbs(name) {
+		return fmt.Errorf("cannot create file with relative path: %q", name)
+	}
+	if strings.HasSuffix(name, "/") {
+		return fmt.Errorf("cannot create non-file path: %q", name)
+	}
```

Comment:
We should be checking the suffix after filepath.Clean() to avoid "/foo/..". Does it make sense to at the top of this function (and secureMkdirAll):
```
if filepath.Clean(name) != name {
   return fmt.Errorf("...")
}
```

Reply:
I think filepath.Clean chops the trailing `/` but I'll double-check.

Reply:
Also changed this code to reject unclean paths. I think it's good to merge now.